pkg/vpcmodel/multiExplainability.go

```diff
@@ -0,0 +1,185 @@
+// (5) src and dst are internal nodes from different vpcs and has no connection:         {src,dst,nil}
+func CreateMultiExplanationsInput(cConfigs MultipleVPCConfigs, conns map[string]*GroupConnLines) []explainInputEntry {
+	internalNodes, externalNodes := collectNodesForExplanation(cConfigs, conns)
+	multiVpcConnections := collectMultiConnectionsForExplanation(cConfigs, conns)
+	explanationsInput := []explainInputEntry{}
+	for src, srcConfig := range internalNodes {
+		for dst, dstConfig := range internalNodes {
+			var vpcConfig *VPCConfig
+			if multiVpcConfig, ok := multiVpcConnections[src][dst]; ok {
+				vpcConfig = multiVpcConfig // input of case (4)
+			} else if srcConfig == dstConfig {
+				vpcConfig = srcConfig // input of case (1)
+			} // else - input of case (5)
+			explanationsInput = append(explanationsInput, explainInputEntry{vpcConfig, src, dst})
+		}
+		for external := range externalNodes {
+			explanationsInput = append(explanationsInput,
+				explainInputEntry{srcConfig, src, external}, // input of case (2)
+				explainInputEntry{srcConfig, external, src}) // input of case (3)
+		}
+	}
+	return explanationsInput
+}
+
+func collectNodesForExplanation(cConfigs MultipleVPCConfigs, conns map[string]*GroupConnLines) (
+	internalNodes map[EndpointElem]*VPCConfig, externalNodes map[EndpointElem]bool) {
+	internalNodes = map[EndpointElem]*VPCConfig{}
+	externalNodes = map[EndpointElem]bool{}
+	for _, vpcConfig := range cConfigs {
+		if !vpcConfig.IsMultipleVPCsConfig {
+			for _, n := range vpcConfig.Nodes {
+				if !n.IsExternal() {
+					internalNodes[n] = vpcConfig
+				}
+			}
+		}
+	}
+	// we collect only external nodes with connections:
+	for _, vpcConn := range conns {
+		for _, line := range vpcConn.GroupedLines {
+			if eSrc, ok := line.src.(*groupedExternalNodes); ok {
+				externalNodes[eSrc] = true
+			}
+			if eDst, ok := line.dst.(*groupedExternalNodes); ok {
+				externalNodes[eDst] = true
+			}
+		}
+	}
+	return internalNodes, externalNodes
+}
+
+func collectMultiConnectionsForExplanation(
+	cConfigs MultipleVPCConfigs, conns map[string]*GroupConnLines) map[EndpointElem]map[EndpointElem]*VPCConfig {
+	multiVpcConnections := map[EndpointElem]map[EndpointElem]*VPCConfig{}
+	for vpcName, vpcConfig := range cConfigs {
+		if vpcConfig.IsMultipleVPCsConfig {
+			vpcConn := conns[vpcName]
+			for _, line := range vpcConn.GroupedLines {
+				srcs := []EndpointElem{line.src}
+				dsts := []EndpointElem{line.dst}
+				if srcList, ok := line.src.(*groupedEndpointsElems); ok {
+					srcs = *srcList
+				}
+				if dstList, ok := line.dst.(*groupedEndpointsElems); ok {
+					dsts = *dstList
+				}
+				for _, src := range srcs {
+					for _, dst := range dsts {
+						if _, ok := multiVpcConnections[src]; !ok {
+							multiVpcConnections[src] = map[EndpointElem]*VPCConfig{}
+						}
+						multiVpcConnections[src][dst] = vpcConfig
+					}
+				}
+			}
```

Comment:
can get the connections between src-dst pairs from different VPCs without the grouping (from `AllowedConnsCombined`), and then no need to iterate src/dst lists per EndpointElem, I think it can be more straight-forward.

Reply:
fixwd